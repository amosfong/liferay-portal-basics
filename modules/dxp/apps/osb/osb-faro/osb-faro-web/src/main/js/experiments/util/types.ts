export type FormatYAxisFn = (metricUnit: string) => (value: number) => string;

export type GetFormattedMedianFn = (
	median: number,
	metric: MetricName
) => string;

export type GetMetricNameFn = (metric: MetricName) => string;

export type GetMetricUnitFn = (metric: MetricName) => string;

export type GetStatusColorFn = (status: Status) => string;

export type GetStatusNameFn = (status: Status) => string;

export type GetShortIntervals = (intervals: Array<Date>) => Array<Date>;

export type GetTicksFn = (maxValue: number) => Array<number>;

export type GetVariantLabels = (experiment: {
	bestVariant?: Variant;
	dxpVariantId: string;
	publishedDXPVariantId?: string;
	status: Status;
	winnerDXPVariantId?: string;
}) => Array<{status: string; value: string}>;

type MetricName =
	| 'BOUNCE_RATE'
	| 'CLICK_RATE'
	| 'MAX_SCROLL_DEPTH'
	| 'TIME_ON_PAGE';

type MergedVariant = {
	changes: number;
	confidenceInterval: Array<number>;
	control: boolean;
	dxpVariantId: string;
	dxpVariantName: string;
	improvement: number;
	median: number;
	probabilityToWin: number;
	trafficSplit: number;
	uniqueVisitors: number;
};

export type MergedVariantsFn = (
	variants: Array<Variant>,
	variantMetrics: Array<VariantMetric>
) => Array<MergedVariant>;

type Status =
	| 'COMPLETED'
	| 'DRAFT'
	| 'FINISHED_NO_WINNER'
	| 'FINISHED_WINNER'
	| 'PAUSED'
	| 'RUNNING'
	| 'SCHEDULED'
	| 'TERMINATED';

type Variant = {
	changes: number;
	control: boolean;
	dxpVariantId: string;
	dxpVariantName: string;
	trafficSplit: number;
	uniqueVisitors: number;
};

type VariantMetric = {
	confidenceInterval: Array<number>;
	dxpVariantId: string;
	improvement: number;
	median: number;
	probabilityToWin: number;
};
