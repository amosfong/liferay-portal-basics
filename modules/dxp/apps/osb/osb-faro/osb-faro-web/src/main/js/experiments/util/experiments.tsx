import {CHART_COLORS} from 'shared/util/charts';
import {CONTROL_COLOR} from './constants';
import {
	FormatYAxisFn,
	GetFormattedMedianFn,
	GetMetricNameFn,
	GetMetricUnitFn,
	GetShortIntervals,
	GetStatusColorFn,
	GetStatusNameFn,
	GetTicksFn,
	GetVariantLabels,
	MergedVariantsFn
} from './types';
import {round} from 'lodash';
import {toRounded, toThousands, toThousandsBase} from 'shared/util/numbers';

const METRICS_NAMES = new Map([
	['BOUNCE_RATE', Liferay.Language.get('bounce-rate')],
	['CLICK_RATE', Liferay.Language.get('click-through-rate')],
	['MAX_SCROLL_DEPTH', Liferay.Language.get('max-scroll-depth')],
	['TIME_ON_PAGE', Liferay.Language.get('view-duration')]
]);

const METRICS_UNITS = new Map([
	['BOUNCE_RATE', '%'],
	['CLICK_RATE', '%'],
	['MAX_SCROLL_DEPTH', '%'],
	['TIME_ON_PAGE', 's']
]);

const STATUS_COLORS = new Map([
	['COMPLETED', 'success'],
	['DRAFT', 'secondary'],
	['FINISHED_NO_WINNER', 'secondary'],
	['FINISHED_WINNER', 'success'],
	['PAUSED', 'secondary'],
	['RUNNING', 'info'],
	['SCHEDULED', 'warning'],
	['TERMINATED', 'danger']
]);

const STATUS_NAMES = new Map([
	['COMPLETED', Liferay.Language.get('complete')],
	['DRAFT', Liferay.Language.get('draft')],
	['FINISHED_NO_WINNER', Liferay.Language.get('no-winner')],
	['FINISHED_WINNER', Liferay.Language.get('winner-declared')],
	['PAUSED', Liferay.Language.get('paused')],
	['RUNNING', Liferay.Language.get('running')],
	['SCHEDULED', Liferay.Language.get('scheduled')],
	['TERMINATED', Liferay.Language.get('terminated')]
]);

const getExperimentLink = ({action, id, pageURL}) => {
	const experimentLink = `${pageURL}?segmentsExperimentKey=${id}`;

	if (action) {
		return `${experimentLink}&segmentsExperimentAction=${action}`;
	}

	return experimentLink;
};

export const formatYAxis: FormatYAxisFn = metricUnit => value => {
	if (value % 1 === 0) {
		return `${value}${metricUnit}`;
	}

	return `${value.toFixed(1)}${metricUnit}`;
};

export const getFormattedMedian: GetFormattedMedianFn = (median, metric) => {
	const precision = metric === 'CLICK_RATE' ? 3 : 2;

	return toRounded(median, precision);
};

export const getFormattedMedianLabel = metric =>
	metric === 'CLICK_RATE'
		? `${Liferay.Language.get('median')} ${getMetricName(metric)}`
		: `${getMetricName(metric)} ${Liferay.Language.get('median')}`;

export const getFormattedProbabilityToWin = value => {
	if (value < 0.1) {
		value = '< 0.1';
	} else if (value > 99.9) {
		value = '> 99.9';
	} else {
		value = toRounded(value);
	}

	return value;
};

export const getMetricName: GetMetricNameFn = metric =>
	METRICS_NAMES.get(metric);

export const getMetricUnit: GetMetricUnitFn = metric =>
	METRICS_UNITS.get(metric);

export const getStatusColor: GetStatusColorFn = status =>
	STATUS_COLORS.get(status);

export const getStatusName: GetStatusNameFn = status =>
	STATUS_NAMES.get(status).toUpperCase();

export const mergedVariants: MergedVariantsFn = (variants, variantMetrics) =>
	variants.map(variant => ({
		...variant,
		...variantMetrics.find(
			({dxpVariantId}) => variant.dxpVariantId === dxpVariantId
		)
	}));

export const getActions = (
	status: string,
	{id, onDelete, pageURL, publishable} = null
) => {
	const deleteButton = {
		displayType: 'secondary',
		label: Liferay.Language.get('delete'),
		onClick: onDelete
	};

	switch (status) {
		case 'COMPLETED': {
			return [deleteButton];
		}
		case 'DRAFT': {
			return [
				{
					displayType: 'primary',
					label: Liferay.Language.get('review'),
					redirectURL: getExperimentLink({
						action: 'reviewAndRun',
						id,
						pageURL
					})
				},
				{
					displayType: 'secondary',
					label: Liferay.Language.get('delete'),
					redirectURL: getExperimentLink({
						action: 'delete',
						id,
						pageURL
					})
				}
			];
		}
		case 'FINISHED_NO_WINNER':
		case 'FINISHED_WINNER': {
			return [
				{
					displayType: 'primary',
					label: Liferay.Language.get('publish'),
					redirectURL: getExperimentLink({
						action: 'publish',
						id,
						pageURL
					})
				},
				{
					displayType: 'secondary',
					label: Liferay.Language.get('delete'),
					redirectURL: getExperimentLink({
						action: 'delete',
						id,
						pageURL
					})
				}
			];
		}
		case 'TERMINATED': {
			if (publishable) {
				return [
					{
						displayType: 'primary',
						label: Liferay.Language.get('publish'),
						redirectURL: getExperimentLink({
							action: 'publish',
							id,
							pageURL
						})
					},
					{
						displayType: 'secondary',
						label: Liferay.Language.get('delete'),
						redirectURL: getExperimentLink({
							action: 'delete',
							id,
							pageURL
						})
					}
				];
			}

			return [deleteButton];
		}
		case 'RUNNING': {
			return [
				{
					displayType: 'secondary',
					label: Liferay.Language.get('terminate'),
					redirectURL: getExperimentLink({
						action: 'terminate',
						id,
						pageURL
					})
				}
			];
		}
		default: {
			return [];
		}
	}
};

export const getBestVariant = ({
	dxpVariants,
	goal,
	metrics: {variantMetrics}
}) => {
	if (
		!dxpVariants ||
		variantMetrics.every(({median}) => median === variantMetrics[0].median)
	) {
		return null;
	}

	if (goal?.metric === 'BOUNCE_RATE') {
		return mergedVariants(
			dxpVariants,
			variantMetrics
		).reduce((prev, current) =>
			prev.median < current.median ? prev : current
		);
	}

	return mergedVariants(dxpVariants, variantMetrics).reduce((prev, current) =>
		prev.median > current.median ? prev : current
	);
};

export const getVariantLabels: GetVariantLabels = ({
	bestVariant,
	dxpVariantId,
	publishedDXPVariantId,
	status,
	winnerDXPVariantId
}) => {
	const labels = [];

	if (status === 'RUNNING' && bestVariant?.dxpVariantId === dxpVariantId) {
		labels.push({
			status: 'success',
			value: Liferay.Language.get('current-best')
		});
	}

	if (
		winnerDXPVariantId === dxpVariantId &&
		(status === 'COMPLETED' || status === 'FINISHED_WINNER')
	) {
		labels.push({
			status: 'success',
			value: Liferay.Language.get('winner')
		});
	}

	if (publishedDXPVariantId === dxpVariantId) {
		labels.push({
			status: 'info',
			value: Liferay.Language.get('published')
		});
	}

	return labels;
};

export const getTicks: GetTicksFn = maxValue => {
	const arr = [];
	let interval = 1;
	const step = Math.round(maxValue / 8);

	while (interval <= maxValue) {
		arr.push(interval);

		interval = interval + step;
	}

	return [...arr];
};

export const getShortIntervals: GetShortIntervals = intervals =>
	getTicks(intervals.length).map(tick => intervals[tick - 1]);

export const toThousandsABTesting = number => {
	if (number > 1e4) {
		return toThousandsBase(number, factor =>
			Math.trunc(round(number * factor, 2))
		);
	}

	return toThousands(number);
};

export const getLegendData = dxpVariants => {
	const COLORS = [...CHART_COLORS];

	return dxpVariants.map(({control, dxpVariantId, dxpVariantName}) => ({
		color: control ? CONTROL_COLOR : COLORS.shift(),
		id: dxpVariantId,
		name: dxpVariantName
	}));
};

export const getMedianGraphData = ({dxpVariants, metricUnit}) => {
	const COLORS = [...CHART_COLORS];

	const type = metricUnit === '%' ? 'percentage' : 'number';

	const formatter =
		metricUnit === '%' ? value => value : value => `${value}s`;

	const items = dxpVariants.map(({confidenceInterval, control, median}) => ({
		intervals: [
			{
				end: confidenceInterval[1],
				start: confidenceInterval[0]
			}
		],
		progress: [
			{
				color: control ? CONTROL_COLOR : COLORS.shift(),
				value: median
			}
		]
	}));

	const maxValue = Math.max(
		...dxpVariants.map(({confidenceInterval}) => confidenceInterval[1])
	);

	return {
		disableScroll: true,
		empty: maxValue === 0,
		formatSpacement: false,
		grid: {
			formatter,
			maxValue,
			minValue: 0,
			precision: 2,
			show: true,
			type
		},
		items
	};
};
