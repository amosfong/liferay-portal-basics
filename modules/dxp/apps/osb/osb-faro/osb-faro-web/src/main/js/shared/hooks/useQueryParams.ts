import {useLocation} from 'react-router-dom';

function decodeQueryParam(param) {
	return decodeURIComponent(param.replace(/\+/g, ' '));
}

function queryStringToObject(initialQueryString: string): any {
	if (!initialQueryString) return {};

	const queryString = initialQueryString.replace('?', '');

	const params = queryString.split('&');
	const query = {};

	params.forEach(param => {
		const [key, value] = param.split('=');
		query[key] = decodeQueryParam(value);
	});

	return query;
}

// TODO: Remove this once we upgrade to react-router-dom v6
export function useQueryParams() {
	const {search} = useLocation();

	return queryStringToObject(search);
}
