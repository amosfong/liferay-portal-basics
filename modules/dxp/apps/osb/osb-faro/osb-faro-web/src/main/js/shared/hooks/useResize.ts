import {MutableRefObject, useEffect, useState} from 'react';

export const useResize = (
	elementRef: MutableRefObject<any> | null
): [number, number] => {
	const [size, setSize] = useState<[number, number]>([0, 0]);

	useEffect(() => {
		const handleResize = () =>
			setSize([
				elementRef?.current?.clientWidth ?? window.innerWidth,
				elementRef?.current?.clientHeight ?? window.innerHeight
			]);

		handleResize();

		window.addEventListener('resize', handleResize);

		return () => window.removeEventListener('resize', handleResize);
	}, [elementRef]);

	return size;
};
