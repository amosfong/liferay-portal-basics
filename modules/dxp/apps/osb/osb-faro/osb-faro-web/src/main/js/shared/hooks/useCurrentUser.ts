import {fetchCurrentUser} from 'shared/actions/users';
import {useDispatch, useSelector} from 'react-redux';
import {useEffect} from 'react';
import {User} from 'shared/util/records';

/**
 * Used only on the first time on App.tsx to fetch data from backend.
 * To get currentUser, you can use WithCurrentUser (for HOC) or useCurrentUser (for HOOK).
 */
export const useFetchCurrentUser = (initialGroupId: string = '0') => {
	const currentUser = useSelector<any, any>(state =>
		state.get('currentUser')
	);
	const error = currentUser.get('error');
	const loading = currentUser.get('loading');

	const data = useCurrentUser();

	const dispatch = useDispatch();

	useEffect(() => {
		let groupId = '0';

		if (initialGroupId && initialGroupId !== 'add') {
			groupId = initialGroupId;
		}

		dispatch(fetchCurrentUser(groupId));
	}, [initialGroupId]);

	return {
		data,
		error,
		loading
	};
};

/**
 * Get currentUser from redux store.
 */
export const useCurrentUser = (): User => {
	const currentUserId = useSelector<any, any>(state =>
		state.getIn(['currentUser', 'data'])
	);
	const data: User = useSelector<any, any>(state =>
		state.getIn(['users', currentUserId, 'data'])
	);

	const newUser = new User({
		emailAddress: '',
		id: '',
		name: '',
		roleName: '',
		status: 1
	});

	return data || newUser;
};
