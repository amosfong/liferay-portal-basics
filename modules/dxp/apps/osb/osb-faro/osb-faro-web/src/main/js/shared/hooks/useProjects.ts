import {fetchJoinableProjects, fetchMany} from 'shared/api/projects';
import {fromJS} from 'immutable';
import {Project} from 'shared/util/records';
import {useEffect, useState} from 'react';

export const useFetchProjects = () => {
	const [loading, setLoading] = useState(true);
	const [data, setData] = useState([]);

	useEffect(() => {
		async function fetch() {
			try {
				const projects = await fetchMany();

				setData(projects.map(result => new Project(fromJS(result))));
				setLoading(false);
			} catch {
				throw new Error('Error on fetchProjects');
			}
		}

		fetch();
	}, []);

	return {
		data,
		loading
	};
};

export const useFetchJoinableProjects = () => {
	const [loading, setLoading] = useState(true);
	const [data, setData] = useState([]);

	useEffect(() => {
		async function fetch() {
			try {
				const projects = await fetchJoinableProjects();

				setData(projects);
				setLoading(false);
			} catch {
				throw new Error('Error on fetchJoinableProjects');
			}
		}

		fetch();
	}, []);

	return {
		data,
		loading
	};
};
