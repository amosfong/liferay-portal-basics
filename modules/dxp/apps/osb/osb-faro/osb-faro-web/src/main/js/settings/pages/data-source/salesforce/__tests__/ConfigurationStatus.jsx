import * as data from 'test/data';
import mockStore from 'test/mock-store';
import React from 'react';
import SalesforceConfigurationStatus from '../ConfigurationStatus';
import {BrowserRouter} from 'react-router-dom';
import {DataSource} from 'shared/util/records';
import {fromJS, Map} from 'immutable';
import {Provider} from 'react-redux';
import {render} from '@testing-library/react';
import {useCurrentUser} from 'shared/hooks/useCurrentUser';

jest.unmock('react-dom');

jest.mock('shared/hooks/useCurrentUser', () => ({
	useCurrentUser: jest.fn()
}));

const DEFAULT_DATA_SOURCE = new DataSource(
	data.mockLiferayDataSource(23, {
		connectionStatuses: fromJS({
			accounts: {
				modifiedDate: data.getTimestamp(),
				status: 1
			},
			contacts: {
				modifiedDate: data.getTimestamp(),
				status: 1
			},
			leads: {
				modifiedDate: data.getTimestamp(),
				status: 1
			}
		}),
		provider: new Map({
			accountsConfiguration: {},
			contactsConfiguration: {},
			leadsConfiguration: {},
			type: 'SALESFORCE'
		})
	})
);

const WrappedComponent = ({dataSource = DEFAULT_DATA_SOURCE}) => (
	<Provider store={mockStore()}>
		<BrowserRouter>
			<SalesforceConfigurationStatus
				dataSource={dataSource}
				groupId='23'
				id='23'
			/>
		</BrowserRouter>
	</Provider>
);

describe('SalesforceConfigurationStatus', () => {
	it('should render', () => {
		useCurrentUser.mockImplementation(() => ({isAdmin: () => true}));

		const {container} = render(<WrappedComponent />);

		expect(container).toMatchSnapshot();
	});

	it('should render as read-only if the user is not authorized to make changes', () => {
		useCurrentUser.mockImplementation(() => ({isAdmin: () => false}));

		const {container} = render(<WrappedComponent />);

		expect(container).toMatchSnapshot();
	});

	// TODO: should be rewritten when new endpoint to replace connectionStatuses is added
	it('should render already configured items with the button text "pause" instead of "enable"', () => {
		useCurrentUser.mockImplementation(() => ({isAdmin: () => true}));

		const {container} = render(
			<WrappedComponent
				dataSource={
					new DataSource(
						data.mockLiferayDataSource(23, {
							connectionStatuses: fromJS({
								accounts: {
									modifiedDate: data.getTimestamp(),
									status: 2
								},
								contacts: {
									modifiedDate: data.getTimestamp(),
									status: 2
								},
								leads: {
									modifiedDate: data.getTimestamp(),
									status: 2
								}
							}),
							provider: new Map({
								accountsConfiguration: {},
								contactsConfiguration: {},
								leadsConfiguration: {},
								type: 'SALESFORCE'
							})
						})
					)
				}
			/>
		);

		expect(container).toMatchSnapshot();
	});

	// TODO: should be rewritten when new endpoint to replace connectionStatuses is added
	it('should render as configuring items with a metric bar instead of a button', () => {
		useCurrentUser.mockImplementation(() => ({isAdmin: () => true}));

		const {container} = render(
			<WrappedComponent
				dataSource={
					new DataSource(
						data.mockLiferayDataSource(23, {
							connectionStatuses: fromJS({
								accounts: {
									count: 123,
									modifiedDate: data.getTimestamp(),
									status: 1,
									totalCount: 1234
								},
								contacts: {
									count: 573,
									modifiedDate: data.getTimestamp(),
									status: 1,
									totalCount: 5433
								},
								leads: {
									count: 573,
									modifiedDate: data.getTimestamp(),
									status: 1,
									totalCount: 5433
								}
							}),
							provider: new Map({
								type: 'SALESFORCE'
							})
						})
					)
				}
			/>
		);

		expect(container).toMatchSnapshot();
	});
});
