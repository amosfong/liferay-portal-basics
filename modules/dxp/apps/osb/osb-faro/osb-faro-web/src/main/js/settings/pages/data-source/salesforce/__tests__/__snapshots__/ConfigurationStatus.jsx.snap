// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SalesforceConfigurationStatus should render 1`] = `
<div>
  <div
    class="settings-base-page-root data-source-base-page-root"
  >
    <div
      class="maintenance-alert-root"
    />
    <ol
      class="breadcrumb breadcrumbs-root"
    >
      <li
        class="breadcrumb-item"
      >
        <a
          class="breadcrumb-text-truncate breadcrumb-link"
          href="/workspace/23/settings/data-source"
        >
          Data Sources
        </a>
      </li>
      <li
        class="breadcrumb-item"
      >
        <button
          class="breadcrumb-text-truncate btn btn-unstyled"
          type="button"
        >
          Liferay23
        </button>
      </li>
      <li
        class="breadcrumb-item active"
      >
        <span
          class="breadcrumb-text-truncate"
        >
          Configuration Status
        </span>
      </li>
    </ol>
    <div
      class="content-header"
    >
      <div
        class="header-text"
      >
        <div
          class="d-flex"
        >
          <h3
            class="title-text text-truncate"
          >
            <span
              class="text-truncate"
              data-tooltip="false"
              title=""
            >
              Configure Salesforce
            </span>
          </h3>
        </div>
        <div
          class="description"
        >
          Liferay23 - https://www.faro-23.io
        </div>
      </div>
      <div
        class="page-actions-container"
      />
    </div>
    <div>
      <div
        class="page-container"
      >
        <div
          class="content-main"
        >
          <div
            class="sheet data-source-base-tabs-root page-display"
          >
            <div
              class="sheet-body"
            >
              <ul
                class="nav-root nav nav-underline"
                role="tablist"
              >
                <li
                  class="nav-item"
                  role="presentation"
                >
                  <a
                    class="nav-link"
                    href="/workspace/23/settings/data-source/23"
                    role="tab"
                  >
                    <div
                      class="h4"
                    >
                      Authorization
                    </div>
                  </a>
                </li>
                <li
                  class="nav-item active"
                  role="presentation"
                >
                  <a
                    class="nav-link active"
                    href="/workspace/23/settings/data-source/23/salesforce/configuration-status"
                    role="tab"
                  >
                    <div
                      class="h4"
                    >
                      Configure Data Source
                    </div>
                  </a>
                </li>
              </ul>
            </div>
            <div
              class="sheet-body"
            >
              <div
                class="configuration-overview-root"
              >
                <h3
                  class="sheet-subtitle"
                >
                  Data Configuration
                </h3>
                <div
                  class="base-configuration-item-root"
                >
                  <div
                    class="actions"
                  >
                    <a
                      class="button-root btn btn-block btn-secondary"
                      display="secondary"
                      href="/workspace/23/settings/data-source/23/salesforce/field-mapping/accounts"
                    >
                      View
                    </a>
                  </div>
                  <div
                    class="info"
                  >
                    <div
                      class="title"
                    >
                      <b>
                        Accounts
                      </b>
                    </div>
                    <div
                      class="text-secondary"
                    >
                      <div
                        class="description"
                      >
                        Represents an organization or person involved with your business.
                      </div>
                      
                    </div>
                  </div>
                </div>
                <div
                  class="base-configuration-item-root"
                >
                  <div
                    class="actions"
                  >
                    <a
                      class="button-root btn btn-block btn-secondary"
                      display="secondary"
                      href="/workspace/23/settings/data-source/23/salesforce/field-mapping/individuals"
                    >
                      View
                    </a>
                  </div>
                  <div
                    class="info"
                  >
                    <div
                      class="title"
                    >
                      <b>
                        Individuals
                      </b>
                    </div>
                    <div
                      class="text-secondary"
                    >
                      <div
                        class="description"
                      >
                        Represented by a contact or lead within Salesforce.
                      </div>
                      
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="content-side"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`SalesforceConfigurationStatus should render already configured items with the button text "pause" instead of "enable" 1`] = `
<div>
  <div
    class="settings-base-page-root data-source-base-page-root"
  >
    <div
      class="maintenance-alert-root"
    />
    <ol
      class="breadcrumb breadcrumbs-root"
    >
      <li
        class="breadcrumb-item"
      >
        <a
          class="breadcrumb-text-truncate breadcrumb-link"
          href="/workspace/23/settings/data-source"
        >
          Data Sources
        </a>
      </li>
      <li
        class="breadcrumb-item"
      >
        <button
          class="breadcrumb-text-truncate btn btn-unstyled"
          type="button"
        >
          Liferay23
        </button>
      </li>
      <li
        class="breadcrumb-item active"
      >
        <span
          class="breadcrumb-text-truncate"
        >
          Configuration Status
        </span>
      </li>
    </ol>
    <div
      class="content-header"
    >
      <div
        class="header-text"
      >
        <div
          class="d-flex"
        >
          <h3
            class="title-text text-truncate"
          >
            <span
              class="text-truncate"
              data-tooltip="false"
              title=""
            >
              Configure Salesforce
            </span>
          </h3>
        </div>
        <div
          class="description"
        >
          Liferay23 - https://www.faro-23.io
        </div>
      </div>
      <div
        class="page-actions-container"
      />
    </div>
    <div>
      <div
        class="page-container"
      >
        <div
          class="content-main"
        >
          <div
            class="sheet data-source-base-tabs-root page-display"
          >
            <div
              class="sheet-body"
            >
              <ul
                class="nav-root nav nav-underline"
                role="tablist"
              >
                <li
                  class="nav-item"
                  role="presentation"
                >
                  <a
                    class="nav-link"
                    href="/workspace/23/settings/data-source/23"
                    role="tab"
                  >
                    <div
                      class="h4"
                    >
                      Authorization
                    </div>
                  </a>
                </li>
                <li
                  class="nav-item active"
                  role="presentation"
                >
                  <a
                    class="nav-link active"
                    href="/workspace/23/settings/data-source/23/salesforce/configuration-status"
                    role="tab"
                  >
                    <div
                      class="h4"
                    >
                      Configure Data Source
                    </div>
                  </a>
                </li>
              </ul>
            </div>
            <div
              class="sheet-body"
            >
              <div
                class="configuration-overview-root"
              >
                <h3
                  class="sheet-subtitle"
                >
                  Data Configuration
                </h3>
                <div
                  class="base-configuration-item-root"
                >
                  <div
                    class="actions"
                  >
                    <a
                      class="button-root btn btn-block btn-secondary"
                      display="secondary"
                      href="/workspace/23/settings/data-source/23/salesforce/field-mapping/accounts"
                    >
                      View
                    </a>
                  </div>
                  <div
                    class="info"
                  >
                    <div
                      class="title"
                    >
                      <b>
                        Accounts
                      </b>
                    </div>
                    <div
                      class="text-secondary"
                    >
                      <div
                        class="description"
                      >
                        Represents an organization or person involved with your business.
                      </div>
                      
                    </div>
                  </div>
                </div>
                <div
                  class="base-configuration-item-root"
                >
                  <div
                    class="actions"
                  >
                    <a
                      class="button-root btn btn-block btn-secondary"
                      display="secondary"
                      href="/workspace/23/settings/data-source/23/salesforce/field-mapping/individuals"
                    >
                      View
                    </a>
                  </div>
                  <div
                    class="info"
                  >
                    <div
                      class="title"
                    >
                      <b>
                        Individuals
                      </b>
                    </div>
                    <div
                      class="text-secondary"
                    >
                      <div
                        class="description"
                      >
                        Represented by a contact or lead within Salesforce.
                      </div>
                      
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="content-side"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`SalesforceConfigurationStatus should render as configuring items with a metric bar instead of a button 1`] = `
<div>
  <div
    class="settings-base-page-root data-source-base-page-root"
  >
    <div
      class="maintenance-alert-root"
    />
    <ol
      class="breadcrumb breadcrumbs-root"
    >
      <li
        class="breadcrumb-item"
      >
        <a
          class="breadcrumb-text-truncate breadcrumb-link"
          href="/workspace/23/settings/data-source"
        >
          Data Sources
        </a>
      </li>
      <li
        class="breadcrumb-item"
      >
        <button
          class="breadcrumb-text-truncate btn btn-unstyled"
          type="button"
        >
          Liferay23
        </button>
      </li>
      <li
        class="breadcrumb-item active"
      >
        <span
          class="breadcrumb-text-truncate"
        >
          Configuration Status
        </span>
      </li>
    </ol>
    <div
      class="content-header"
    >
      <div
        class="header-text"
      >
        <div
          class="d-flex"
        >
          <h3
            class="title-text text-truncate"
          >
            <span
              class="text-truncate"
              data-tooltip="false"
              title=""
            >
              Configure Salesforce
            </span>
          </h3>
        </div>
        <div
          class="description"
        >
          Liferay23 - https://www.faro-23.io
        </div>
      </div>
      <div
        class="page-actions-container"
      />
    </div>
    <div>
      <div
        class="page-container"
      >
        <div
          class="content-main"
        >
          <div
            class="sheet data-source-base-tabs-root page-display"
          >
            <div
              class="sheet-body"
            >
              <ul
                class="nav-root nav nav-underline"
                role="tablist"
              >
                <li
                  class="nav-item"
                  role="presentation"
                >
                  <a
                    class="nav-link"
                    href="/workspace/23/settings/data-source/23"
                    role="tab"
                  >
                    <div
                      class="h4"
                    >
                      Authorization
                    </div>
                  </a>
                </li>
                <li
                  class="nav-item active"
                  role="presentation"
                >
                  <a
                    class="nav-link active"
                    href="/workspace/23/settings/data-source/23/salesforce/configuration-status"
                    role="tab"
                  >
                    <div
                      class="h4"
                    >
                      Configure Data Source
                    </div>
                    <span
                      class="badge badge-primary ml-2"
                    >
                      <span
                        class="badge-item badge-item-expand"
                      >
                        2
                      </span>
                    </span>
                  </a>
                </li>
              </ul>
            </div>
            <div
              class="sheet-body"
            >
              <div
                class="configuration-overview-root"
              >
                <h3
                  class="sheet-subtitle"
                >
                  Data Configuration
                </h3>
                <div
                  class="base-configuration-item-root"
                >
                  <div
                    class="actions"
                  >
                    <a
                      class="button-root btn btn-block btn-secondary"
                      display="secondary"
                      href="/workspace/23/settings/data-source/23/salesforce/field-mapping/accounts"
                    >
                      View
                    </a>
                  </div>
                  <div
                    class="info"
                  >
                    <div
                      class="title"
                    >
                      <b>
                        Accounts
                      </b>
                    </div>
                    <div
                      class="text-secondary"
                    >
                      <div
                        class="description"
                      >
                        Represents an organization or person involved with your business.
                      </div>
                      
                    </div>
                  </div>
                </div>
                <div
                  class="base-configuration-item-root"
                >
                  <div
                    class="actions"
                  >
                    <a
                      class="button-root btn btn-block btn-secondary"
                      display="secondary"
                      href="/workspace/23/settings/data-source/23/salesforce/field-mapping/individuals"
                    >
                      View
                    </a>
                  </div>
                  <div
                    class="info"
                  >
                    <div
                      class="title"
                    >
                      <b>
                        Individuals
                      </b>
                    </div>
                    <div
                      class="text-secondary"
                    >
                      <div
                        class="description"
                      >
                        Represented by a contact or lead within Salesforce.
                      </div>
                      
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="content-side"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`SalesforceConfigurationStatus should render as read-only if the user is not authorized to make changes 1`] = `
<div>
  <div
    class="settings-base-page-root data-source-base-page-root"
  >
    <div
      class="maintenance-alert-root"
    />
    <ol
      class="breadcrumb breadcrumbs-root"
    >
      <li
        class="breadcrumb-item"
      >
        <a
          class="breadcrumb-text-truncate breadcrumb-link"
          href="/workspace/23/settings/data-source"
        >
          Data Sources
        </a>
      </li>
      <li
        class="breadcrumb-item"
      >
        <button
          class="breadcrumb-text-truncate btn btn-unstyled"
          type="button"
        >
          Liferay23
        </button>
      </li>
      <li
        class="breadcrumb-item active"
      >
        <span
          class="breadcrumb-text-truncate"
        >
          Configuration Status
        </span>
      </li>
    </ol>
    <div
      class="content-header"
    >
      <div
        class="header-text"
      >
        <div
          class="d-flex"
        >
          <h3
            class="title-text text-truncate"
          >
            <span
              class="text-truncate"
              data-tooltip="false"
              title=""
            >
              Configure Salesforce
            </span>
          </h3>
        </div>
        <div
          class="description"
        >
          Liferay23 - https://www.faro-23.io
        </div>
      </div>
      <div
        class="page-actions-container"
      />
    </div>
    <div>
      <div
        class="page-container"
      >
        <div
          class="content-main"
        >
          <div
            class="sheet data-source-base-tabs-root page-display"
          >
            <div
              class="sheet-body"
            >
              <ul
                class="nav-root nav nav-underline"
                role="tablist"
              >
                <li
                  class="nav-item"
                  role="presentation"
                >
                  <a
                    class="nav-link"
                    href="/workspace/23/settings/data-source/23"
                    role="tab"
                  >
                    <div
                      class="h4"
                    >
                      Authorization
                    </div>
                  </a>
                </li>
                <li
                  class="nav-item active"
                  role="presentation"
                >
                  <a
                    class="nav-link active"
                    href="/workspace/23/settings/data-source/23/salesforce/configuration-status"
                    role="tab"
                  >
                    <div
                      class="h4"
                    >
                      Configure Data Source
                    </div>
                  </a>
                </li>
              </ul>
            </div>
            <div
              class="sheet-body"
            >
              <div
                class="configuration-overview-root"
              >
                <h3
                  class="sheet-subtitle"
                >
                  Data Configuration
                </h3>
                <div
                  class="base-configuration-item-root"
                >
                  <div
                    class="actions"
                  >
                    <a
                      class="button-root link-disabled btn btn-block btn-secondary"
                      display="secondary"
                    >
                      View
                    </a>
                  </div>
                  <div
                    class="info"
                  >
                    <div
                      class="title"
                    >
                      <b>
                        Accounts
                      </b>
                    </div>
                    <div
                      class="text-secondary"
                    >
                      <div
                        class="description"
                      >
                        Represents an organization or person involved with your business.
                      </div>
                      
                    </div>
                  </div>
                </div>
                <div
                  class="base-configuration-item-root"
                >
                  <div
                    class="actions"
                  >
                    <a
                      class="button-root link-disabled btn btn-block btn-secondary"
                      display="secondary"
                    >
                      View
                    </a>
                  </div>
                  <div
                    class="info"
                  >
                    <div
                      class="title"
                    >
                      <b>
                        Individuals
                      </b>
                    </div>
                    <div
                      class="text-secondary"
                    >
                      <div
                        class="description"
                      >
                        Represented by a contact or lead within Salesforce.
                      </div>
                      
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="content-side"
        />
      </div>
    </div>
  </div>
</div>
`;
